src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -1650,40 +1652,97 @@ BOOST_AUTO_TEST_CASE(GradientConcatenateLayerTest)
  */
 BOOST_AUTO_TEST_CASE(SimpleLookupLayerTest)
 {
-  arma::mat output, input, delta, gradient;
-  Lookup<> module(10, 5);
+  const size_t vocabSize = 10;
+  const size_t embeddingSize = 2;
+  const size_t seqLength = 3;
+  const size_t batchSize = 4;
+
+  arma::mat output, input, gy, g, gradient;
+
+  Lookup<> module(vocabSize, embeddingSize);
   module.Parameters().randu();
 
   // Test the Forward function.
-  input = arma::zeros(2, 1);
-  input(0) = 1;
-  input(1) = 3;
+  input = arma::zeros(seqLength, batchSize);
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+    int token = math::RandInt(1, vocabSize);
+    input(i) = token;
+  }
 
   module.Forward(input, output);
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    // The Lookup module uses index - 1 for the cols.
+    const double outputSum = arma::accu(module.Parameters().cols(
+        arma::conv_to<arma::uvec>::from(input.col(i)) - 1));
 
-  // The Lookup module uses index - 1 for the cols.
-  const double outputSum = arma::accu(module.Parameters().col(0)) +
-      arma::accu(module.Parameters().col(2));
-
-  BOOST_REQUIRE_CLOSE(outputSum, arma::accu(output), 1e-3);
+    BOOST_REQUIRE_CLOSE(outputSum, arma::accu(output.col(i)), 1e-3);
+  }
 
   // Test the Backward function.
-  module.Backward(input, input, delta);
-  BOOST_REQUIRE_EQUAL(arma::accu(input), arma::accu(input));
+  gy = 0.3 * arma::randu(embeddingSize * seqLength, batchSize);
+  module.Backward(input, gy, g);
+  BOOST_REQUIRE_EQUAL(arma::accu(gy), arma::accu(g));
 
   // Test the Gradient function.
-  arma::mat error = arma::ones(2, 5);
-  error = error.t();
-  error.col(1) *= 0.5;
-
+  arma::mat error = 0.01 * arma::randu(embeddingSize * seqLength, batchSize);
   module.Gradient(input, error, gradient);
 
-  // The Lookup module uses index - 1 for the cols.
-  const double gradientSum = arma::accu(gradient.col(0)) +
-      arma::accu(gradient.col(2));
+  BOOST_CHECK_CLOSE_FRACTION(arma::accu(error), arma::accu(gradient), 1e-05);
```

Comment:
Basically, you pick some random columns from `error` several times (they may even repeat since `input` is also random) and sum them up. That's how you obtain `gradient`. Then you compare the sum of `error` and the sum of `gradient`. I don't think they should be equal. The numbers are random. They could have the same expectation if you divided them by the number of summands.

Note, the previous test dealt with deterministic numbers. Both `input` and `error` were defined. So, it could have sense. I don't think the present version is correct.

Reply:
Random numbers are not a problem. Maybe `print` statements can reveal it to you.
Here is the output:
```cpp
input:
   2.0000   2.0000   2.0000   9.0000
   8.0000   3.0000   9.0000   7.0000
   9.0000   5.0000   9.0000   9.0000
error:
   0.0003   0.0048   0.0071   0.0042
   0.0015   0.0030   0.0072   0.0060
   0.0098   0.0029   0.0007   0.0009
   0.0062   0.0092   0.0049   0.0027
   0.0030   0.0063   0.0009   0.0062
   0.0036   0.0076   0.0076   0.0022
error in 3D form:
[cube slice 0]
   0.0003   0.0098   0.0030
   0.0015   0.0062   0.0036

[cube slice 1]
   0.0048   0.0029   0.0063
   0.0030   0.0092   0.0076

[cube slice 2]
   0.0071   0.0007   0.0009
   0.0072   0.0049   0.0076

[cube slice 3]
   0.0042   0.0009   0.0062
   0.0060   0.0027   0.0022

gradient initially:
        0        0        0        0        0        0        0        0        0        0
        0        0        0        0        0        0        0        0        0        0
gradient after evaluation of 1th batch
        0   0.0003        0        0        0        0        0   0.0098   0.0030        0
        0   0.0015        0        0        0        0        0   0.0062   0.0036        0
gradient after evaluation of 2th batch
        0   0.0051   0.0029        0   0.0063        0        0   0.0098   0.0030        0
        0   0.0045   0.0092        0   0.0076        0        0   0.0062   0.0036        0
gradient after evaluation of 3th batch
        0   0.0123   0.0029        0   0.0063        0        0   0.0098   0.0046        0
        0   0.0117   0.0092        0   0.0076        0        0   0.0062   0.0161        0
gradient after evaluation of 4th batch
        0   0.0123   0.0029        0   0.0063        0   0.0009   0.0098   0.0151        0
        0   0.0117   0.0092        0   0.0076        0   0.0027   0.0062   0.0243        0
```

Now let's focus on 1st column of input, 1st slice of error in 3D form, and gradient after evaluation of 1st batch.
You can see that the first column of input has (2, 8, 9). Those are the tokens for the words in that batch. Now columns 2, 8, 9 is selected from gradient and then 1st batch (1st slice) of error is added. So what happened is-- 
`gradient.cols(2, 8, 9) += error.slice(0)`. See that LHS and RHS have same shapes, so it's a valid operation.

...and same thing goes for every batch.
Hope this helped :)

Reply:
Thanks for the clarification! I totally agree. I don't know why, but I thought you select random columns from `error`. In fact you select random columns from `gradient` and add a slice of `error` to the selected columns. Of course in this case the test is absolutely correct.